benchmarks/src/main/java/org/opensearch/benchmark/routing/allocation/AllocationBenchmark.java

```diff
@@ -71,75 +73,103 @@
     // support to constrain the combinations of benchmark parameters and we do not want to rely on OptionsBuilder as each benchmark would
     // need its own main method and we cannot execute more than one class with a main method per JAR.
     @Param({
-        // indices| shards| replicas| nodes
-        "       10|      1|        0|     1",
-        "       10|      3|        0|     1",
-        "       10|     10|        0|     1",
-        "      100|      1|        0|     1",
-        "      100|      3|        0|     1",
-        "      100|     10|        0|     1",
-
-        "       10|      1|        0|    10",
-        "       10|      3|        0|    10",
-        "       10|     10|        0|    10",
-        "      100|      1|        0|    10",
-        "      100|      3|        0|    10",
-        "      100|     10|        0|    10",
-
-        "       10|      1|        1|    10",
-        "       10|      3|        1|    10",
-        "       10|     10|        1|    10",
-        "      100|      1|        1|    10",
-        "      100|      3|        1|    10",
-        "      100|     10|        1|    10",
-
-        "       10|      1|        2|    10",
-        "       10|      3|        2|    10",
-        "       10|     10|        2|    10",
-        "      100|      1|        2|    10",
-        "      100|      3|        2|    10",
-        "      100|     10|        2|    10",
-
-        "       10|      1|        0|    50",
-        "       10|      3|        0|    50",
-        "       10|     10|        0|    50",
-        "      100|      1|        0|    50",
-        "      100|      3|        0|    50",
-        "      100|     10|        0|    50",
-
-        "       10|      1|        1|    50",
-        "       10|      3|        1|    50",
-        "       10|     10|        1|    50",
-        "      100|      1|        1|    50",
-        "      100|      3|        1|    50",
-        "      100|     10|        1|    50",
-
-        "       10|      1|        2|    50",
-        "       10|      3|        2|    50",
-        "       10|     10|        2|    50",
-        "      100|      1|        2|    50",
-        "      100|      3|        2|    50",
-        "      100|     10|        2|    50" })
-    public String indicesShardsReplicasNodes = "10|1|0|1";
+        // indices| shards| replicas| source| target| concurrentRecoveries
```

Comment:
Can you run the benchmark to publish the results

Reply:
I have started the benchmark run. Will publish results as and when they are available

Reply:
Benchmark is still running for higher configurations, publishing whatever we have so far:

```
# Benchmark: 
# Parameters: (indicesShardsReplicasSourceTargetRecoveries =        10|      2|        0|       1|      1|      1|)
  5.521 ±(99.9%) 0.648 ms/op [Average]
  (min, avg, max) = (4.172, 5.521, 8.551), stdev = 0.970
  CI (99.9%): [4.873, 6.170] (assumes normal distribution)

# Parameters: (indicesShardsReplicasSourceTargetRecoveries =        10|      3|        0|       1|      1|      2|)
  5.787 ±(99.9%) 0.619 ms/op [Average]
  (min, avg, max) = (4.317, 5.787, 7.938), stdev = 0.926
  CI (99.9%): [5.169, 6.406] (assumes normal distribution)

# Parameters: (indicesShardsReplicasSourceTargetRecoveries =        10|     10|        0|       1|      1|      5|)
  5.865 ±(99.9%) 0.652 ms/op [Average]
  (min, avg, max) = (4.199, 5.865, 8.339), stdev = 0.976
  CI (99.9%): [5.213, 6.517] (assumes normal distribution)

# Parameters: (indicesShardsReplicasSourceTargetRecoveries =       100|      1|        0|       1|      1|     10|)
  6.240 ±(99.9%) 0.671 ms/op [Average]
  (min, avg, max) = (3.926, 6.240, 8.353), stdev = 1.004
  CI (99.9%): [5.569, 6.911] (assumes normal distribution)

# Parameters: (indicesShardsReplicasSourceTargetRecoveries =       100|      3|        0|       1|      1|     10|)
  6.585 ±(99.9%) 0.600 ms/op [Average]
  (min, avg, max) = (5.334, 6.585, 8.456), stdev = 0.899
  CI (99.9%): [5.984, 7.185] (assumes normal distribution)

# Parameters: (indicesShardsReplicasSourceTargetRecoveries =       100|     10|        0|       1|      1|     10|)
  7.531 ±(99.9%) 0.441 ms/op [Average]
  (min, avg, max) = (6.647, 7.531, 10.135), stdev = 0.660
  CI (99.9%): [7.089, 7.972] (assumes normal distribution)

# Parameters: (indicesShardsReplicasSourceTargetRecoveries =        10|      2|        0|      10|     10|      1|)
  5.969 ±(99.9%) 0.633 ms/op [Average]
  (min, avg, max) = (4.596, 5.969, 8.979), stdev = 0.947
  CI (99.9%): [5.336, 6.602] (assumes normal distribution)

# Parameters: (indicesShardsReplicasSourceTargetRecoveries =        10|      3|        0|      10|      5|      2|)
  5.905 ±(99.9%) 0.540 ms/op [Average]
  (min, avg, max) = (4.471, 5.905, 8.141), stdev = 0.808
  CI (99.9%): [5.365, 6.445] (assumes normal distribution)
```

Reply:
Thanks Ankit, unfortunately we might need benchmarks w and w/o the change to see the actual gain

Reply:
Please find the benchmark numbers below:

```
benchmark                                       indices         shards          replica         src nodes       tgt nodes      conc recoveries  non-optimized score             optimized score         percent improvement
measureExclusionOnZoneAwareStartedShard         10              2               0               1               1               1               5.864 ± 0.625 ms/op             5.521 ± 0.648 ms/op             5.85
measureExclusionOnZoneAwareStartedShard         10              3               0               1               1               2               5.999 ± 0.448 ms/op             5.787 ± 0.619 ms/op             3.53
measureExclusionOnZoneAwareStartedShard         10              10              0               1               1               5               5.983 ± 0.485 ms/op             5.865 ± 0.652 ms/op             1.97
measureExclusionOnZoneAwareStartedShard         100             1               0               1               1               10              6.582 ± 0.696 ms/op             6.240 ± 0.671 ms/op             5.20
measureExclusionOnZoneAwareStartedShard         100             3               0               1               1               10              7.060 ± 0.452 ms/op             6.585 ± 0.600 ms/op             6.73
measureExclusionOnZoneAwareStartedShard         100             10              0               1               1               10              10.423 ± 0.800 ms/op            7.531 ± 0.441 ms/op             27.75
measureExclusionOnZoneAwareStartedShard         10              2               0               10              10              1               5.925 ± 0.478 ms/op             5.969 ± 0.633 ms/op             -0.74
measureExclusionOnZoneAwareStartedShard         10              3               0               10              5               2               5.715 ± 0.370 ms/op             5.905 ± 0.540 ms/op             -3.32
measureExclusionOnZoneAwareStartedShard         10              10              0               10              5               5               7.308 ± 0.418 ms/op             5.890 ± 0.472 ms/op             19.40
measureExclusionOnZoneAwareStartedShard         100             1               0               5               10              5               8.305 ± 0.909 ms/op             6.190 ± 0.565 ms/op             25.47
measureExclusionOnZoneAwareStartedShard         100             3               0               10              5               5               11.674 ± 0.716 ms/op            6.957 ± 0.670 ms/op             40.41
measureExclusionOnZoneAwareStartedShard         100             10              0               10              20              6               53.977 ± 2.654 ms/op            9.266 ± 0.829 ms/op             82.83
measureExclusionOnZoneAwareStartedShard         10              1               1               10              10              1               5.590 ± 0.480 ms/op             6.050 ± 0.668 ms/op             -8.23
measureExclusionOnZoneAwareStartedShard         10              3               1               10              3               3               6.810 ± 0.594 ms/op             5.670 ± 0.545 ms/op             16.74
measureExclusionOnZoneAwareStartedShard         10              10              1               5               12              5               11.508 ± 0.728 ms/op            6.375 ± 0.705 ms/op             44.60
measureExclusionOnZoneAwareStartedShard         100             1               1               10              10              6               10.845 ± 0.910 ms/op            7.591 ± 0.546 ms/op             30.00
measureExclusionOnZoneAwareStartedShard         100             3               1               10              5               8               18.840 ± 1.287 ms/op            7.398 ± 0.687 ms/op             60.73
measureExclusionOnZoneAwareStartedShard         100             10              1               8               17              8               105.254 ± 6.125 ms/op           10.694 ± 0.757 ms/op            89.84
measureExclusionOnZoneAwareStartedShard         10              1               2               10              10              1               6.145 ± 0.555 ms/op             5.961 ± 0.642 ms/op             2.99
measureExclusionOnZoneAwareStartedShard         10              3               2               10              5               3               7.331 ± 0.495 ms/op             6.462 ± 0.701 ms/op             11.85
measureExclusionOnZoneAwareStartedShard         10              10              2               5               10              5               14.740 ± 0.918 ms/op            6.624 ± 0.526 ms/op             55.06
measureExclusionOnZoneAwareStartedShard         100             1               2               10              8               7               14.360 ± 1.230 ms/op            8.006 ± 0.893 ms/op             44.25
measureExclusionOnZoneAwareStartedShard         100             3               2               13              17              5               56.852 ± 3.348 ms/op            9.279 ± 1.312 ms/op             83.68
measureExclusionOnZoneAwareStartedShard         100             10              2               10              20              8               186.101 ± 9.656 ms/op           14.557 ± 1.058 ms/op            92.18
measureExclusionOnZoneAwareStartedShard         10              2               1               20              20              1               6.994 ± 0.413 ms/op             6.507 ± 0.576 ms/op             6.96
measureExclusionOnZoneAwareStartedShard         10              3               1               20              30              1               9.654 ± 0.717 ms/op             5.818 ± 0.405 ms/op             39.73
measureExclusionOnZoneAwareStartedShard         10              10              1               20              10              3               12.820 ± 1.401 ms/op            6.544 ± 0.590 ms/op             48.95
measureExclusionOnZoneAwareStartedShard         100             1               1               20              5               5               11.600 ± 0.993 ms/op            6.772 ± 0.826 ms/op             41.62
measureExclusionOnZoneAwareStartedShard         100             3               1               20              23              6               45.449 ± 3.944 ms/op            9.566 ± 0.779 ms/op             78.95
measureExclusionOnZoneAwareStartedShard         100             10              1               40              20              8               158.230 ± 7.729 ms/op           14.070 ± 1.124 ms/op            91.11
measureExclusionOnZoneAwareStartedShard         10              3               2               50              30              1               13.371 ± 0.846 ms/op            6.976 ± 0.621 ms/op             47.83
measureExclusionOnZoneAwareStartedShard         10              3               2               50              25              1               13.561 ± 0.809 ms/op            6.222 ± 0.702 ms/op             54.12
measureExclusionOnZoneAwareStartedShard         10              10              1               50              33              2               23.353 ± 0.815 ms/op            8.505 ± 0.845 ms/op             63.58
measureExclusionOnZoneAwareStartedShard         100             1               1               40              50              2               25.773 ± 1.710 ms/op            8.052 ± 0.532 ms/op             68.76
measureExclusionOnZoneAwareStartedShard         100             3               1               50              70              3               108.306 ± 7.946 ms/op           12.299 ± 0.803 ms/op            88.64
measureExclusionOnZoneAwareStartedShard         100             10              1               60              50              3               347.034 ± 30.735 ms/op          16.265 ± 1.207 ms/op            95.31
measureExclusionOnZoneAwareStartedShard         10              10              2               50              50              1               50.281 ± 3.700 ms/op            8.396 ± 0.796 ms/op             83.30
measureExclusionOnZoneAwareStartedShard         10              3               2               50              30              1               13.992 ± 1.138 ms/op            6.892 ± 0.616 ms/op             50.74
measureExclusionOnZoneAwareStartedShard         10              10              2               50              40              2               39.764 ± 2.001 ms/op            8.754 ± 0.778 ms/op             77.99
measureExclusionOnZoneAwareStartedShard         100             1               2               40              50              2               45.836 ± 3.938 ms/op            9.316 ± 0.879 ms/op             79.68
measureExclusionOnZoneAwareStartedShard         100             3               2               50              30              6               103.342 ± 9.791 ms/op           13.797 ± 0.992 ms/op            86.65
measureExclusionOnZoneAwareStartedShard         100             10              2               33              55              6               480.793 ± 25.636 ms/op          20.216 ± 1.513 ms/op            95.80
measureExclusionOnZoneAwareStartedShard         500             60              1               100             100             12              19269.360 ± 344.191 ms/op       329.574 ± 12.378 ms/op          98.29
measureExclusionOnZoneAwareStartedShard         500             60              1               100             40              12              11336.688 ± 276.501 ms/op       285.680 ± 12.235 ms/op          97.48
measureExclusionOnZoneAwareStartedShard         500             60              1               40              100             12              16267.428 ± 357.485 ms/op       275.860 ± 9.164 ms/op           98.30
measureExclusionOnZoneAwareStartedShard         50              60              1               100             100             6               1736.322 ± 69.013 ms/op         55.793 ± 2.555 ms/op            96.79
measureExclusionOnZoneAwareStartedShard         50              60              1               100             40              6               1068.012 ± 37.326 ms/op         31.407 ± 2.524 ms/op            97.06
measureExclusionOnZoneAwareStartedShard         50              60              1               40              100             6               1542.845 ± 43.455 ms/op         29.182 ± 1.776 ms/op            98.11

```

Reply:
```
benchmark                               indices         shards          replica         src nodes       tgt nodes       conc recoveries non-optimized score             optimized score         percent improvement
measureShardRelocationComplete          10              2               0               1               1               1               6.103 ± 0.637 ms/op             6.197 ± 0.552 ms/op             -1.54
measureShardRelocationComplete          10              3               0               1               1               2               6.588 ± 0.743 ms/op             5.988 ± 0.573 ms/op             9.11
measureShardRelocationComplete          10              10              0               1               1               5               6.877 ± 0.689 ms/op             6.772 ± 0.781 ms/op             1.53
measureShardRelocationComplete          100             1               0               1               1               10              6.995 ± 0.549 ms/op             6.580 ± 0.530 ms/op             5.93
measureShardRelocationComplete          100             3               0               1               1               10              8.823 ± 0.862 ms/op             8.250 ± 0.926 ms/op             6.49
measureShardRelocationComplete          100             10              0               1               1               10              14.772 ± 1.423 ms/op            11.499 ± 1.007 ms/op            22.16
measureShardRelocationComplete          10              2               0               10              10              1               6.168 ± 0.516 ms/op             6.338 ± 0.605 ms/op             -2.76
measureShardRelocationComplete          10              3               0               10              5               2               6.591 ± 0.552 ms/op             6.338 ± 0.609 ms/op             3.84
measureShardRelocationComplete          10              10              0               10              5               5               8.140 ± 0.924 ms/op             6.874 ± 0.613 ms/op             15.55
measureShardRelocationComplete          100             1               0               5               10              5               9.588 ± 0.553 ms/op             6.930 ± 0.879 ms/op             27.72
measureShardRelocationComplete          100             3               0               10              5               5               14.255 ± 1.817 ms/op            8.309 ± 0.570 ms/op             41.71
measureShardRelocationComplete          100             10              0               10              20              6               66.269 ± 4.465 ms/op            14.578 ± 1.130 ms/op            78.00
measureShardRelocationComplete          10              1               1               10              10              1               6.590 ± 0.802 ms/op             6.820 ± 0.578 ms/op             -3.49
measureShardRelocationComplete          10              3               1               10              3               3               7.810 ± 0.657 ms/op             7.299 ± 1.185 ms/op             6.54
measureShardRelocationComplete          10              10              1               5               12              5               13.058 ± 1.318 ms/op            7.218 ± 0.656 ms/op             44.72
measureShardRelocationComplete          100             1               1               10              10              6               13.314 ± 1.054 ms/op            8.851 ± 1.168 ms/op             33.52
measureShardRelocationComplete          100             3               1               10              5               8               22.452 ± 1.930 ms/op            10.175 ± 0.580 ms/op            54.68
measureShardRelocationComplete          100             10              1               8               17              8               115.146 ± 9.760 ms/op           22.330 ± 3.422 ms/op            80.61
measureShardRelocationComplete          10              1               2               10              10              1               6.341 ± 0.620 ms/op             6.245 ± 0.601 ms/op             1.51
measureShardRelocationComplete          10              3               2               10              5               3               8.025 ± 0.655 ms/op             6.365 ± 0.548 ms/op             20.69
measureShardRelocationComplete          10              10              2               5               10              5               15.870 ± 1.105 ms/op            7.880 ± 0.692 ms/op             50.35
measureShardRelocationComplete          100             1               2               10              8               7               16.143 ± 1.221 ms/op            9.559 ± 0.603 ms/op             40.79
measureShardRelocationComplete          100             3               2               13              17              5               60.807 ± 4.183 ms/op            12.526 ± 1.142 ms/op            79.40
measureShardRelocationComplete          100             10              2               10              20              8               194.576 ± 12.283 ms/op          24.863 ± 1.888 ms/op            87.22
measureShardRelocationComplete          10              2               1               20              20              1               7.753 ± 0.717 ms/op             6.946 ± 0.564 ms/op             10.41
measureShardRelocationComplete          10              3               1               20              30              1               10.368 ± 0.907 ms/op            6.264 ± 0.497 ms/op             39.58
measureShardRelocationComplete          10              10              1               20              10              3               14.260 ± 1.301 ms/op            7.670 ± 0.566 ms/op             46.21
measureShardRelocationComplete          100             1               1               20              5               5               13.422 ± 0.837 ms/op            7.454 ± 0.907 ms/op             44.46
measureShardRelocationComplete          100             3               1               20              23              6               48.509 ± 2.972 ms/op            12.982 ± 1.009 ms/op            73.24
measureShardRelocationComplete          100             10              1               40              20              8               169.456 ± 6.631 ms/op           25.495 ± 3.150 ms/op            84.95
measureShardRelocationComplete          10              3               2               50              30              1               15.016 ± 1.425 ms/op            7.211 ± 0.584 ms/op             51.98
measureShardRelocationComplete          10              3               2               50              25              1               14.158 ± 1.151 ms/op            7.444 ± 0.579 ms/op             47.42
measureShardRelocationComplete          10              10              1               50              33              2               25.819 ± 2.031 ms/op            8.824 ± 0.763 ms/op             65.82
measureShardRelocationComplete          100             1               1               40              50              2               29.462 ± 2.101 ms/op            10.496 ± 0.856 ms/op            64.37
measureShardRelocationComplete          100             3               1               50              70              3               112.412 ± 7.976 ms/op           16.176 ± 1.497 ms/op            85.61
measureShardRelocationComplete          100             10              1               60              50              3               352.369 ± 21.597 ms/op          25.972 ± 2.007 ms/op            92.63
measureShardRelocationComplete          10              10              2               50              50              1               54.177 ± 4.423 ms/op            9.773 ± 0.865 ms/op             81.96
measureShardRelocationComplete          10              3               2               50              30              1               15.184 ± 1.262 ms/op            7.677 ± 0.598 ms/op             49.44
measureShardRelocationComplete          10              10              2               50              40              2               45.177 ± 3.614 ms/op            10.294 ± 0.727 ms/op            77.21
measureShardRelocationComplete          100             1               2               40              50              2               48.868 ± 3.604 ms/op            11.003 ± 1.000 ms/op            77.48
measureShardRelocationComplete          100             3               2               50              30              6               108.770 ± 9.062 ms/op           19.605 ± 1.999 ms/op            81.98
measureShardRelocationComplete          100             10              2               33              55              6               534.918 ± 33.208 ms/op          34.530 ± 2.552 ms/op            93.54
measureShardRelocationComplete          500             60              1               100             100             12              19617.938 ± 397.195 ms/op       768.885 ± 39.174 ms/op          96.08
measureShardRelocationComplete          500             60              1               100             40              12              11823.170 ± 245.410 ms/op       733.790 ± 64.055 ms/op          93.79
measureShardRelocationComplete          500             60              1               40              100             12              16706.173 ± 330.388 ms/op       624.669 ± 28.434 ms/op          96.26
measureShardRelocationComplete          50              60              1               100             100             6               1736.313 ± 72.725 ms/op         85.427 ± 5.528 ms/op            95.08
measureShardRelocationComplete          50              60              1               100             40              6               1093.102 ± 62.901 ms/op         59.250 ± 4.551 ms/op            94.58
measureShardRelocationComplete          50              60              1               40              100             6               1558.825 ± 78.488 ms/op         57.933 ± 5.569 ms/op            96.28
```

Reply:
My conclusion from the benchmark is no regression with significant improvement for certain cases